tensorstore/kvstore/s3/s3_key_value_store.cc

```diff
@@ -0,0 +1,1306 @@
+    auto request = HttpRequestBuilder("HEAD", url).BuildRequest();
+    auto future = driver->transport_->IssueRequest(request, {});
+    if(!future.status().ok()) return future.status();
+    auto & headers = future.value().headers;
+    if(auto it = headers.find(kAmzBucketRegionHeader); it !=headers.end()) {
+      driver->aws_region_ = it->second;
+      driver->endpoint_ = tensorstore::StrCat(
+        "https://", data_.bucket, ".s3.", driver->aws_region_, kDotAmazonAwsDotCom);
+    } else {
+      return absl::InvalidArgumentError(
+        tensorstore::StrCat("bucket ", data_.bucket, " does not exist"));
+    }
+  } else {
+    auto endpoint = data_.endpoint.value();
+    auto parsed = internal::ParseGenericUri(endpoint);
+    if(parsed.scheme != "http" && parsed.scheme != "https") {
+      return absl::InvalidArgumentError(
+        tensorstore::StrCat("Endpoint ", endpoint,
+                            " has invalid schema ", parsed.scheme,
+                            ". Should be http(s)."));
+    }
+    if(!parsed.query.empty()) {
+      return absl::InvalidArgumentError(
+        tensorstore::StrCat("Query in endpoint unsupported ", endpoint));
+    }
+    if(!parsed.fragment.empty()) {
+      return absl::InvalidArgumentError(
+        tensorstore::StrCat("Fragment in endpoint unsupported ", endpoint));
+    }
+
+    driver->aws_region_ = data_.aws_region;
+    driver->endpoint_ = endpoint;
+  }
+
+  ABSL_LOG(INFO) << "S3 driver using endpoint [" << driver->endpoint_ << "]";
+
+  auto parsed = internal::ParseGenericUri(driver->endpoint_);
+  size_t end_of_host = parsed.authority_and_path.find('/');
+  driver->host_ = parsed.authority_and_path.substr(0, end_of_host);
+
+  // NOTE: Remove temporary logging use of experimental feature.
+  if (data_.rate_limiter.has_value()) {
+    ABSL_LOG(INFO) << "Using experimental_s3_rate_limiter";
+  }
+  return driver;
+}
+
+/// A ReadTask is a function object used to satisfy a
+/// S3KeyValueStore::Read request.
+struct ReadTask : public RateLimiterNode,
+                  public internal::AtomicReferenceCount<ReadTask> {
+  IntrusivePtr<S3KeyValueStore> owner;
+  std::string read_url;
+  kvstore::ReadOptions options;
+  Promise<kvstore::ReadResult> promise;
+
+  int attempt_ = 0;
+  absl::Time start_time_;
+
+  ReadTask(IntrusivePtr<S3KeyValueStore> owner, std::string read_url,
+           kvstore::ReadOptions options, Promise<kvstore::ReadResult> promise)
+      : owner(std::move(owner)),
+        read_url(std::move(read_url)),
+        options(std::move(options)),
+        promise(std::move(promise)) {}
+
+  ~ReadTask() { owner->admission_queue().Finish(this); }
+
+  static void Start(void* task) {
+    auto* self = reinterpret_cast<ReadTask*>(task);
+    self->owner->read_rate_limiter().Finish(self);
+    self->owner->admission_queue().Admit(self, &ReadTask::Admit);
+  }
+
+  static void Admit(void* task) {
```

Comment:
my naming for this is perhaps a bit confusing. I should probably use OnAdmit or something less mentally colliding.

Reply:
I think  the `ReadTask` type qualifies it well enough for comprehension.  I guess `OnXXX` might distinguish callbacks from other methods on the struct. I don't feel strongly about this so I haven't changed anything here. Let me know if you think differently.